src/pmu.rs

```diff
@@ -0,0 +1,297 @@
+    0x1C0, // repeats
+    0x1C0, 0x1C0, 0x1C0,
+];
+
+#[cfg(not(feature = "g002"))]
+const DEFAULT_WAKE_PROGRAM: [u32; 8] = [
+    0x1F0, // assert all resets and enable all power supplies
+    0x0F8, // idle 2^8 cycles, then deassert hfclkrst
+    0x030, // deassert corerst and padrst
+    0x030, // repeats
+    0x030, 0x030, 0x030, 0x030,
+];
+
+// Hifive1-revB programs
+#[cfg(feature = "g002")]
+const DEFAULT_SLEEP_PROGRAM: [u32; 8] = [
+    0x2F0, // assert corerst
+    0x3F0, // assert hfclkrst
+    0x3D0, // deassert pmu_out_1
+    0x3C0, // deassert pmu_out_0
+    0x3C0, // repeats
+    0x3C0, 0x3C0, 0x3C0,
+];
+
+#[cfg(feature = "g002")]
+const DEFAULT_WAKE_PROGRAM: [u32; 8] = [
+    0x3F0, // assert all resets and enable all power supplies
+    0x2F8, // idle 2^8 cycles, then deassert hfclkrst
+    0x030, // deassert corerst and padrst
+    0x030, // repeats
+    0x030, 0x030, 0x030, 0x030,
+];
+
+///
+/// Enumeration of device reset causes
+/// 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ResetCause {
+    /// Reset due to power on
+    PowerOn,
+    /// Reset due to external input (button)
+    External,
+    /// Reset due to watchdog
+    WatchDog,
+}
+
+///
+/// Enumeration of device wakeup causes
+///
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum WakeupCause {
+    /// Wake up due to reset (see ResetCause)
+    Reset(ResetCause),
+    /// Wake up due to RTC clock
+    RTC,
+    /// Wake up due to digital input (button)
+    Digital,
+}
+
+///
+/// Errors for user data backup procedures
+/// 
+#[derive(Debug)]
+pub enum BackupError {
+    /// Emitted when user data is larger than backup registers capacity
+    DataTooLarge, 
+    /// Emitted when user data is not aligned to 4 bytes or more
+    DataMisaligned,
+}
+
+///
+/// Wakeup/Reset cause errors
+/// 
+#[derive(Debug)]
+pub enum CauseError {
+    InvalidCause,
+}
+
+pub trait PMUExt {
+    ///
+    /// Resets SLEEP and WAKE programs on the PMU to defaults
+    ///
+    fn load_default_programs(&self);
+
+    ///
+    /// Puts device to sleep for N seconds, allowing wake-up button to wake it up as well
+    ///
+    /// # Arguments
+    /// 
+    /// *sleep_time* - the amount of time to sleep for in seconds
+    ///
+    /// # Notes
+    /// 
+    /// - enables RTC to be always on
+    /// - sets RTC scale to 1/s
+    ///
+    fn sleep(self, sleep_time: u32);
+
+    ///
+    /// Stores user data `UD` to backup registers.
+    ///
+    /// # Arguments
+    ///
+    /// * `user_data` - the user data to store. *MUST* have alignment of at least 4 and fit into the backup registers
+    ///
+    /// # Returns
+    /// 
+    /// * `Result<UD, BackupError>` - the stored `user_data` is returned on success
+    /// 
+    /// # Errors
+    ///
+    /// * `BackupError::DataTooLarge` - returned if `user_data` cannot fit into backup registers
+    /// * `BackupError::DataMisaligned` - returned if `user_data` is not aligned to at least 4 bytes
```

Comment:
The check seems to be only for *the size of* `user_data`, not its address.

Reply:
Not sure I understand, do you mean it should be worded as `returned if value of user_data cannot fit ...` ?

Reply:
I think you make two assumptions:
- `user_data`'s *address* is 4-aligned
- `user_data`'s *size* is 4-aligned
Might want to document them both.

Reply:
should be fixed now. I don't really assume on user_data's address since that's mostly the compiler's job, I just need the size to be divisible by 4 bytes